pkg/engine/executor/parse_json.go

```diff
@@ -0,0 +1,243 @@
+package executor
+
+import (
+	"bytes"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/apache/arrow-go/v18/arrow"
+	"github.com/apache/arrow-go/v18/arrow/array"
+	"github.com/apache/arrow-go/v18/arrow/memory"
+	"github.com/buger/jsonparser"
+	"github.com/grafana/loki/v3/pkg/engine/internal/types"
+)
+
+const (
+	jsonSpacer = '_'
+)
+
+var (
+	trueBytes  = []byte("true")
+	falseBytes = []byte("false")
+
+	// the rune error replacement is rejected by Prometheus hence replacing them with space.
+	removeInvalidUtf = func(r rune) rune {
+		if r == utf8.RuneError {
+			return 32 // rune value for space
+		}
+		return r
+	}
+)
+
+func BuildJSONColumns(input *array.String, requestedKeys []string, allocator memory.Allocator) ([]string, []arrow.Array) {
+	return buildColumns(input, requestedKeys, allocator, parseJSONLine, types.JSONParserErrorType)
+}
+
+// parseJSONLine parses a single JSON line and extracts key-value pairs
+// implements ParseFunc
+func parseJSONLine(line string, requestedKeys []string) (map[string]string, error) {
+	// Use the refactored JSONParser for nested object handling and number conversion
+	parser := NewJSONParser()
+	return parser.Process(unsafeBytes(line), requestedKeys)
+}
+
+type JSONParser struct {
```

Comment:
This is a port of the v1 [JSONParser](https://github.com/grafana/loki/blob/20e04a4192fe1f8eacebc4ac241a3f293d328e1b/pkg/logql/log/parser.go#L55). The reason for porting is the v1 parser depends on labels and label builders, it it felt worthwhile to me to remove that dependency from the arrow record / column paradigm. I started with a copy paste of that parser and it's test, and tried to change as little as possible to remove the labels dependency.